feat(gamm): CalcJoinPoolShares and CalcExitPoolCoinsFromShares queries

RusAkh opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
v    If your PR doesn't close an issue, that's OK!  Just remove the Closes: #XXX line!
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #2956

## What is the purpose of the change

Adds two queries: 

1 JoinSwapExactAmountIn allows you to query the amount of shares you get by providing X tokens
2 ExitSwapShareAmountIn allows you to query the amount of tokens you get when exiting pool with X shares

## Testing and Verifying

added two tests

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? yes
  - How is the feature or change documented? changelog

RusAkh commented:
@georgemc98 I want to clarify one thing about the second query. Currently it works by equally splitting shares among all tokens in pool and then returns respectful amount of tokens of specific denom. My question is: should it work like this or should it transform all shares just into one specific denom?

georgemc98 commented:
> @georgemc98 I want to clarify one thing about the second query. Currently it works by equally splitting shares among all tokens in pool and then returns respectful amount of tokens of specific denom. My question is: should it work like this or should it transform all shares just into one specific denom?

This is more difficult than the other queries.  The query estimation will be inaccurate since the liquidity is never taken out before the other token in the pool is swapped for the request token denom.  This is what we have so far: https://github.com/apollodao/osmosis/tree/dev/pool-shares-rpc-queries

georgemc98 commented:
> @georgemc98 I want to clarify one thing about the second query. Currently it works by equally splitting shares among all tokens in pool and then returns respectful amount of tokens of specific denom. My question is: should it work like this or should it transform all shares just into one specific denom?

Could I push my `QueryExitPool` message into your branch?

RusAkh commented:
> > @georgemc98 I want to clarify one thing about the second query. Currently it works by equally splitting shares among all tokens in pool and then returns respectful amount of tokens of specific denom. My question is: should it work like this or should it transform all shares just into one specific denom?
> 
> Could I push my `QueryExitPool` message into your branch?

Sure

RusAkh commented:
> > @georgemc98 I want to clarify one thing about the second query. Currently it works by equally splitting shares among all tokens in pool and then returns respectful amount of tokens of specific denom. My question is: should it work like this or should it transform all shares just into one specific denom?
> 
> This is more difficult than the other queries. The query estimation will be inaccurate since the liquidity is never taken out before the other token in the pool is swapped for the request token denom. This is what we have so far: https://github.com/apollodao/osmosis/tree/dev/pool-shares-rpc-queries
> 
> I have a query for `QueryExitPool` and I'm debating if we even need the `QueryExitSwapShareAmountIn` now.

So what do you think, we leave it or remove it for good?

georgemc98 commented:
> > > @georgemc98 I want to clarify one thing about the second query. Currently it works by equally splitting shares among all tokens in pool and then returns respectful amount of tokens of specific denom. My question is: should it work like this or should it transform all shares just into one specific denom?
> > 
> > 
> > This is more difficult than the other queries. The query estimation will be inaccurate since the liquidity is never taken out before the other token in the pool is swapped for the request token denom. This is what we have so far: https://github.com/apollodao/osmosis/tree/dev/pool-shares-rpc-queries
> > I have a query for `QueryExitPool` and I'm debating if we even need the `QueryExitSwapShareAmountIn` now.
> 
> So what do you think, we leave it or remove it for good?

I recommend that we add two endpoints, JoinPool and ExitPool.  JoinPool can take a single asset, or both assets.  Check out my branch where I implement `ExitPool` and `JoinPool`.  Your `JoinSwapExactAmountIn` query is identical to my `JoinPool` query.  I think it would be better to name it `JoinPool` for simplicity's sake.    I cannot push to your branch, so could you allow me access please?

RusAkh commented:
@georgemc98 invited you as a collaborator, please accept

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,65 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
 }
 
+// JoinSwapExactAmountIn queries the amount of shares you get by providing specific amount of tokens
+func (q Querier) JoinSwapExactAmountIn(ctx context.Context, req *types.QueryJoinSwapExactAmountInRequest) (*types.QueryJoinSwapExactAmountInResponse, error) {
```
Can we call this `CalcJoinPoolShares`?

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,65 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+	}, nil
+}
+
+func (q Querier) ExitSwapShareAmountIn(ctx context.Context, req *types.QueryExitSwapShareAmountInRequest) (*types.QueryExitSwapShareAmountInResponse, error) {
```
godoc please

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,65 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+
+	totalSharesAmount := pool.GetTotalShares()
+	if req.ShareInAmount.GTE(totalSharesAmount) || req.ShareInAmount.LTE(sdk.ZeroInt()) {
+		return &types.QueryExitSwapShareAmountInResponse{TokenOutAmount: 0}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "share ratio is zero or negative")
```
please return nil

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,65 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+
+	exitCoins, err := pool.CalcExitPoolCoinsFromShares(sdkCtx, *req.ShareInAmount, exitFee)
+	if err != nil {
+		return &types.QueryExitSwapShareAmountInResponse{TokenOutAmount: 0}, err
```
ditto

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,65 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	pool, err := q.Keeper.GetPoolAndPoke(sdkCtx, req.PoolId)
+	if err != nil {
+		return &types.QueryExitSwapShareAmountInResponse{TokenOutAmount: 0}, err
```
nil please

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,65 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+			return &types.QueryExitSwapShareAmountInResponse{TokenOutAmount: coin.Amount.Uint64()}, nil
+		}
+	}
+	return &types.QueryExitSwapShareAmountInResponse{}, nil
```
Should we return error if denom is not found?

p0mvn commented on x/gamm/keeper/grpc_query_test.go:
```diff
@@ -9,6 +9,32 @@ import (
+		TokenOutDenom: "foo",
+		ShareInAmount: &defaultShares,
+	})
+	suite.Require().NotNil(out)
```
please add table driven tests to cover all edge cases

p0mvn commented on x/gamm/keeper/grpc_query_test.go:
```diff
@@ -9,6 +9,32 @@ import (
+	suite.Require().NotNil(out)
+	suite.Require().NoError(err)
+}
+func (suite *KeeperTestSuite) TestJoinSwapExactAmountIn() {
```
```suggestion

func (suite *KeeperTestSuite) TestJoinSwapExactAmountIn() {
```

p0mvn commented on x/gamm/keeper/grpc_query_test.go:
```diff
@@ -9,6 +9,32 @@ import (
+		TokensIn: sdk.NewCoins(sdk.NewCoin("foo", sdk.NewInt(500))),
+	})
+	suite.Require().NotNil(out)
+	suite.Require().NoError(err)
```
Table driven tests with edge cases please

p0mvn commented:
The discussed approach of having fewer queries that take more inputs makes sense. I would like to suggest considering naming queries w/o "Join/Exit" but with "Calc". We've been trying to use the former abstraction to imply mutations. Since queries are non-mutative, "Calc" abstraction is preferable

RusAkh commented:
@georgemc98 I fixed some things right now but not finished. basically what is left is to finish with tests. If you decide to work on this issue please feel free to push to my branch (recall that I added you to collaborators) thank you!

georgemc98 commented:
> 

I will add table driven tests for `TestCalcJoinPoolShares` today.

georgemc98 commented:
> @georgemc98 I fixed some things right now but not finished. basically what is left is to finish with tests. If you decide to work on this issue please feel free to push to my branch (recall that I added you to collaborators) thank you!

I will add table driven tests for `TestCalcJoinPoolShares` today.

RusAkh commented:
> > @georgemc98 I fixed some things right now but not finished. basically what is left is to finish with tests. If you decide to work on this issue please feel free to push to my branch (recall that I added you to collaborators) thank you!
> 
> I will add table driven tests for `TestCalcJoinPoolShares` today.

Thank you very much! I will finish TestCalcExitPoolCoinsFromShares as soon as I can

georgemc98 commented:
> > > @georgemc98 I fixed some things right now but not finished. basically what is left is to finish with tests. If you decide to work on this issue please feel free to push to my branch (recall that I added you to collaborators) thank you!
> > 
> > 
> > I will add table driven tests for `TestCalcJoinPoolShares` today.
> 
> Thank you very much! I will finish TestCalcExitPoolCoinsFromShares as soon as I can

Sounds good, I pushed my `CalcExitPoolCoinsFromShares` test earlier

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,67 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+	if err != nil {
+		return nil, types.ErrPoolNotFound
+	}
+	denom_liquidity := q.Keeper.GetDenomLiquidity(sdkCtx, req.TokenOutDenom)
```
let's use camelCase please

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,67 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+		return nil, types.ErrPoolNotFound
+	}
+	denom_liquidity := q.Keeper.GetDenomLiquidity(sdkCtx, req.TokenOutDenom)
+	if denom_liquidity == sdk.NewInt(0) {
```
```suggestion
	if denom_liquidity == sdk.ZeroInt() {
```

p0mvn commented on x/gamm/keeper/grpc_query_test.go:
```diff
@@ -4,11 +4,151 @@ import (
+			types.ErrPoolNotFound,
+		},
+		{
+			"token in denom does not exist",
```
consider adding a test case with an invalid `ShareInAmount`

p0mvn commented on x/gamm/keeper/grpc_query_test.go:
```diff
@@ -4,11 +4,151 @@ import (
+		{
+			"valid test case",
+			poolId,
+			"foo",
```
please aim to use `sdk.DefaultBondDenom` or a reusable consant instead of hardcoding the denom name. The reason is it is more difficult to maintain if we keep hard coding more things

RusAkh replied:
unfortunately suite.prepareBalancerPool does not include stake denom, this is the reason I did not use it. If you agree, I can change this method to creatte a pool with stake denom

p0mvn replied:
Makes sense. Let's not block this then

p0mvn commented on x/gamm/types/errors.go:
```diff
@@ -16,18 +16,19 @@ func (e PoolDoesNotExistError) Error() string {
+	ErrInvalidPool          = sdkerrors.Register(ModuleName, 10, "attempting to create an invalid pool")
+	ErrDenomNotFoundInPool  = sdkerrors.Register(ModuleName, 11, "denom does not exist in pool")
+	ErrDenomAlreadyInPool   = sdkerrors.Register(ModuleName, 12, "denom already exists in the pool")
+	ErrInvalidNumberOfCoins = sdkerrors.Register(ModuleName, 13, "balancer pool only supports LP'ing with one asset or all assets in pool")
```
I suggest to consider defining a Go native error instead of SDK error. We are trying to move away from using SDK errors.

georgemc98 replied:
The error is originally a Go native error, but I changed it to match the other tests in the table.  The other tests match SDK errors, and line 161, https://github.com/RusAkh/osmosis-fork/blob/issue2956/x/gamm/keeper/grpc_query_test.go#L160, requires the error to be registered.  The last test in the join pool test table would return the error: `Target error should be in err chain` when returning a Go error: https://github.com/RusAkh/osmosis-fork/blob/issue2956/x/gamm/keeper/grpc_query_test.go#L135.

p0mvn replied:
Sounds good, thanks for clarifying. We will refactor that separately when deprecating skderrors

georgemc98 replied:
> LGTM. I would like one more reviewer to approve this before merging please

Awesome, thank you.

p0mvn reviewed (commented):
Great work. Can we please also rename the PR to account for the API change?

p0mvn reviewed (approved):
LGTM. I would like one more reviewer to approve this before merging please

alexanderbez commented on proto/osmosis/gamm/v1beta1/query.proto:
```diff
@@ -110,6 +121,39 @@ message QueryPoolTypeResponse {
+  ];
+}
+
+//=============================== QueryCalcExitPoolCoinsFromSharesRequest
```
nit: these are not necessary as they're superfluous IMO

apollo-pacman commented:
@p0mvn @alexanderbez what is the process to determine whether these queries can be added to the stargate whitelist or not?

alexanderbez commented:
> @p0mvn @alexanderbez what is the process to determine whether these queries can be added to the stargate whitelist or not?

AFAIK, they have to be deterministic and constant in gas usage.

georgemc98 commented:
> > @p0mvn @alexanderbez what is the process to determine whether these queries can be added to the stargate whitelist or not?
> 
> 
> 
> AFAIK, they have to be deterministic and constant in gas usage.

Is there anything I should do on my end to initiate these queries  being added to the StargateQuery whitelist?

apollo-pacman commented:
> > > @p0mvn @alexanderbez what is the process to determine whether these queries can be added to the stargate whitelist or not?
> > 
> > 
> > AFAIK, they have to be deterministic and constant in gas usage.
> 
> @alexanderbez Is there anything I should do on my end to initiate these queries being added to the StargateQuery whitelist?

@p0mvn @alexanderbez any update on this?

georgemc98 commented:
@alexanderbez @RusAkh I'm seeing there is a merge conflict now in the `x/gamm/types/errors.go` file.  Would you like me to resolve it, or will the Osmosis team fix it?

alexanderbez commented:
Can you resolve the conflicts?

georgemc98 commented:
> Can you resolve the conflicts?

Yes, happy to.

apollo-sturdy commented:
@alexanderbez @p0mvn Should whitelisting of these queries be done in this PR or do we need to create a separate PR for that? Have you guys reviewed the determinism and whatever else is needed to whitelist it?

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,69 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+		return nil, err
+	}
+
+	for _, coin := range exitCoins {
+		if coin.Denom == req.TokenOutDenom {
+			return &types.QueryCalcExitPoolCoinsFromSharesResponse{TokenOutAmount: coin.Amount.Uint64()}, nil
+		}
+	}
```
What is the behavior of multi-asset pools?

Also, what if we want to query for all tokens out and not just one?

Can we cover the above with tests?

RusAkh replied:
working on these issues

RusAkh replied:
covered @p0mvn

p0mvn commented on x/gamm/keeper/grpc_query_test.go:
```diff
@@ -4,11 +4,165 @@ import (
+func (suite *KeeperTestSuite) TestCalcExitPoolCoinsFromShares() {
+	queryClient := suite.queryClient
+	ctx := suite.Ctx
+	poolId := suite.PrepareBalancerPool()
```
Can we add a test with a multi-asset pool?

RusAkh replied:
PrepareBalancerPool creates a pool with 4 assets

georgemc98 replied:
> > @alexanderbez @p0mvn Should whitelisting of these queries be done in this PR or do we need to create a separate PR for that? Have you guys reviewed the determinism and whatever else is needed to whitelist it?
> 
> This looks good to me to whitelist. However, let's do a separate PR to get more eyes from the rest of the team.

Excellent.  I made a new branch based off this branch and created the PR here: https://github.com/osmosis-labs/osmosis/pull/3217.  Please let me know if there is anything else i need to do to get them whitelisted.

p0mvn commented on x/gamm/pool-models/balancer/pool.go:
```diff
@@ -725,7 +725,7 @@ func (p *Pool) CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee s
 		return numShares, tokensJoined, nil
 	} else if tokensIn.Len() != p.NumAssets() {
-		return sdk.ZeroInt(), sdk.NewCoins(), errors.New("balancer pool only supports LP'ing with one asset or all assets in pool")
+		return sdk.ZeroInt(), sdk.NewCoins(), types.ErrInvalidNumberOfCoins
```
This actually causes a state-break. I would prefer that we keep this as a Go error please

p0mvn commented:
> @alexanderbez @p0mvn Should whitelisting of these queries be done in this PR or do we need to create a separate PR for that? Have you guys reviewed the determinism and whatever else is needed to whitelist it?

This looks good to me to whitelist. However, let's do a separate PR to get more eyes from the rest of the team.

RusAkh commented:
@p0mvn changes addressed:
- now takes a []string with denoms and returns sdk.Coins

georgemc98 commented:
> > @alexanderbez @p0mvn Should whitelisting of these queries be done in this PR or do we need to create a separate PR for that? Have you guys reviewed the determinism and whatever else is needed to whitelist it?
> 
> This looks good to me to whitelist. However, let's do a separate PR to get more eyes from the rest of the team.

Excellent. I made a new branch based off this branch and created the PR here: https://github.com/osmosis-labs/osmosis/pull/3217. Please let me know if there is anything else i need to do to get them whitelisted.

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,72 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+	}, nil
+}
+
+// CalcExitPoolCoinsFromShares queries the amount of tokens you get by exiting specific amouint of shares
```
```suggestion
// CalcExitPoolCoinsFromShares queries the number of tokens you get by exiting a specific amount of shares.
// Only returns the coins requested by `TokensOutDenoms`. If `TokensOutDenoms` is empty, nothing is returned
// despite potentially having exiting coins.
```

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,72 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+	}
+
+	coinsOut := sdk.NewCoins()
+	for _, req_denom := range req.TokensOutDenoms {
```
I'm slightly confused about the purpose of `TokensOutDenoms` for filtering. Is this a requirement for this query?

Why can't we let clients filter out the denoms that they want? I think that filtering out by this input here is likely to lead to erroneous use where clients provide only one `TokensOutDenoms` while there are more existing

georgemc98 replied:
Agreed, let me remove that parameter and remove this for loop.

georgemc98 replied:
> I'm slightly confused about the purpose of `TokensOutDenoms` for filtering. Is this a requirement for this query?
> 
> Why can't we let clients filter out the denoms that they want? I think that filtering out by this input here is likely to lead to erroneous use where clients provide only one `TokensOutDenoms` while there are more existing

I've removed this parameter and fixed the typos that you also suggested.

p0mvn commented on x/gamm/keeper/grpc_query_test.go:
```diff
@@ -2,13 +2,201 @@ package keeper_test
+				exitCoins, err := pool.CalcExitPoolCoinsFromShares(ctx, tc.shareInAmount, exitFee)
+				suite.Require().NoError(err)
+
+				// For each coin in exitCoins we are looking for a match in out response
```
```suggestion
				// For each coin in exitCoins we are looking for a match in our response
```

p0mvn reviewed (commented):
Looks good to me after discussing the filtering

apollo-sturdy commented:
@georgemc98 @p0mvn @RusAkh
I noticed that in `JoinPoolNoSwap` it first does:
```go
neededLpLiquidity, err := getMaximalNoSwapLPAmount(ctx, pool, shareOutAmount)
```
before passing this result to:
```go
sharesOut, err = pool.JoinPoolNoSwap(ctx, neededLpLiquidity, pool.GetSwapFee(ctx))
```
which internally calls `CalcJoinPoolNoSwapShares` which this new query also calls.

`getMaximalNoSwapLPAmount` seems to calculate the reverse of `CalcJoinPoolNoSwapShares`, so in theory the use case of being able to use this query to get `sharesOut` and pass this value to `JoinPoolNoSwap` should work, but have we verified that there are no rounding issues or similar that would lead to issues?

For example, let's say I have some amounts of the assets in a pool, and use this new query `QueryCalcJoinPoolSharesRequest` to receive a `share_out_amount` and a `tokens_out` (should maybe be called `tokens_used`...?).
Now I call `MsgJoinPool` with arguments `ShareOutAmount` = `share_out_amount` returned from the query and `TokenInMaxs` = `tokens_out` returned from the query, will this work without erroring due to rounding differences between `CalcJoinPoolNoSwapShares` and `getMaximalNoSwapLPAmount`?

apollo-sturdy commented:
> For example, let's say I have some amounts of the assets in a pool, and use this new query `QueryCalcJoinPoolSharesRequest` to receive a `share_out_amount` and a `tokens_out` (should maybe be called `tokens_used`...?). Now I call `MsgJoinPool` with arguments `ShareOutAmount` = `share_out_amount` returned from the query and `TokenInMaxs` = `tokens_out` returned from the query, will this work without erroring due to rounding differences between `CalcJoinPoolNoSwapShares` and `getMaximalNoSwapLPAmount`?

To expand on this, the reason we would want to pass the exact amounts returned from the query to `MsgJoinPool` is so that we can be sure that exactly this amount of tokens is drawn from our account, so that we can then do a single sided join with the remaining amount. Of course it would be great if this functionality was supported out of the box, but @ValarDragon recommended that we implement it at the smart contract level (https://github.com/osmosis-labs/osmosis/issues/2994#issuecomment-1289240340), so for that to work this query needs to be an exact simulation.

apollo-sturdy commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,72 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+		return nil, err
+	}
+
+	numShares, newLiquidity, err := pool.CalcJoinPoolShares(sdkCtx, req.TokensIn, pool.GetSwapFee(sdkCtx))
```
Ah I see now looking closer at `CalcJoinPoolShares`. Well in that case we would need two queries. As per the original issue https://github.com/osmosis-labs/osmosis/issues/2956 what we actually want is the ability to simulate each type of liquidity provision (`MsgJoinPool` and `MsgJoinSwapExternAmountIn`). These are the only use cases I'm aware of or can think of anyone wanting.

It looks like for `MsgJoinSwapExternAmountIn` it only really does `CalcJoinPoolShares`, so the current query that this PR introduces should work as an exact simulation of `MsgJoinSwapExternAmountIn`?

For `MsgJoinPool` as mentioned it seems to do more than just `CalcJoinPoolNoSwapShares`. Would you be able to help with introducing an exact simulation query for this function?

p0mvn replied:
> For MsgJoinPool as mentioned it seems to do more than just CalcJoinPoolNoSwapShares. Would you be able to help with introducing an exact simulation query for this function?

I think we should be able to do the following:
- create a new query `CalcJoinPoolNoSwapShares` similar to the one introduced in this PR
- add [this logic](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/keeper/pool_service.go#L79-L82) to get maximal no swap lp amounts
- call pool's [`CalcJoinPoolNoSwapShares`](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/types/pool.go#L79)
- add table-driven unit tests
- ideally, add a functional test that makes sure that this new query behaves exactly the same as [`JoinPoolNoSwap`](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/keeper/pool_service.go#L64)
   * I think this can be also tested contract side so is optional

@RusAkh would you be interested in making a follow-up PR with this new query?

RusAkh replied:
yes, I would like to work on it

p0mvn replied:
Awesome. Please feel free to make a follow-up issue and start working on it

RusAkh replied:
@p0mvn I want to clarify something, why would we include [this logic](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/keeper/pool_service.go#L79-L82)? I see we are simulating pool's CalcJoinPoolNoSwapShares method, right? I want to make sure I understand the task correctly

apollo-sturdy replied:
Sounds great. Thank you @p0mvn and @RusAkh!

@RusAkh the reason to include that logic is that we want an exact simulation of `MsgJoinPool`. @p0mvn perhaps it would be clearer if this new query was called something else? E.g. `QueryJoinPool`, `SimulateJoinPool` or something.

apollo-sturdy replied:
@RusAkh Can you tag me in the PR once you open it? Thanks!

p0mvn replied:
> I want to clarify something, why would we include [this logic](https://github.com/osmosis-labs/osmosis/blob/f5c7125f8cc7e58e0ed7912e62333758a9194870/x/gamm/keeper/pool_service.go#L79-L82)? I see we are simulating pool's CalcJoinPoolNoSwapShares method, right?

> the reason to include that logic is that we want an exact simulation of MsgJoinPool

Yes, that's correct. Based on the requirements from @apollo-sturdy , the new query's API should be:
```
shareOutAmount -> tokensIn, sharesOut
```  

In other words, given  the desired shareOutAmount we want to answer:
- how many tokens do we actually need to put in without performing a swap?
- how many shares would that actually return? (there is small rounding difference from shareOutAmount)

By using `getMaximalNoSwapLPAmount` that you highlighted @RusAkh , we can answer the first question.

By calling pool's `CalcJoinPoolNoSwapShares` and giving it the result of  `getMaximalNoSwapLPAmount`, we can answer second question.

Please let me know if that makes sense.

RusAkh replied:
hello @p0mvn, thanks you, it absolutely makes sense, I thought the point of this query was actually the opposite of what you just said, thanks again!

RusAkh replied:
I think there is a problem in getMaximalNoSwapLPAmount

RusAkh replied:
```
totalSharesAmount := pool.GetTotalShares()
	// shareRatio is the desired number of shares, divided by the total number of
	// shares currently in the pool. It is intended to be used in scenarios where you want
	shareRatio := shareOutAmount.ToDec().QuoInt(totalSharesAmount)
	if shareRatio.LTE(sdk.ZeroDec()) || shareRatio.GT(sdk.OneDec()) {
		return sdk.Coins{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "share ratio is zero or negative")
	}
```

RusAkh replied:
It only checks if share ratio is LTE 0, I think it should also check if it is bigger than one, because in case it is >= 1, you are trying to get more shares than there is available. Wdyt @p0mvn @apollo-sturdy ?

apollo-sturdy replied:
> 

Why should you not be able to provide more liquidity than is already in the pool? The amount of "shares available" is infinite, since they are minted upon liquidity provision?

RusAkh replied:
> > It only checks if share ratio is LTE 0, I think it should also check if it is bigger than one, because in case it is >= 1, you are trying to get more shares than there is available. Wdyt @p0mvn @apollo-sturdy ?
> 
> Why should you not be able to provide more liquidity than is already in the pool? The amount of "shares available" is infinite, since they are minted upon liquidity provision?

indeed, thank you for clarification

RusAkh commented:
hello @p0mvn. Are you sure we need to update sdk version in go.mod in this PR?

p0mvn commented:
> hello @p0mvn. Are you sure we need to update sdk version in go.mod in this PR?

Yes, that should be fine. We use a replace directive for cosmos-sdk so the true version is determined by that. Thanks for checking though

p0mvn commented:
I'm going to merge this. `CalcJoinPoolNoSwapShares` can be addressed in a follow-up PR

p0mvn commented on x/gamm/keeper/grpc_query.go:
```diff
@@ -140,6 +140,59 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	pool, err := q.Keeper.getPoolForSwap(sdkCtx, req.PoolId)
```
For balancer pools, this method may call `PokePool()` which updates pool weights. Is this acceptable?

CC: @ValarDragon

ValarDragon replied:
Yeah this is fine, because:
* this won't get saved
* Even if it did, cosmwasm would revert the result
